src/flows/pipes/Visualization/Controls.tsx

```diff
@@ -42,42 +44,49 @@ const Controls: FC<Props> = ({toggle, visible}) => {
     })
   }
 
-  const selectFn = (fn: string) => {
-    const fns = [...(data.functions || [{name: 'mean'}])]
-    let found = false
-    let fnIdx = fns.findIndex(f => f.name === fn)
-
-    // NOTE: I had to kill it this way because i had poluted my local state
-    while (fnIdx !== -1) {
-      found = true
-      fns.splice(fnIdx, 1)
-      fnIdx = fns.findIndex(f => f.name === fn)
+  const options = useMemo(() => {
+    if (!data.functions || !data.functions.length) {
+      return [{name: 'mean'}].map(f => f.name)
     }
+    return data.functions.map(f => f.name)
+  }, [data.functions])
 
-    if (!found) {
-      fns.push({name: fn})
-    }
+  const selectFn = useCallback(
```

Comment:
any chance this duplicated code could be moved to a central spot?

Reply:
could be.. i know the code is copy pasta, but i'm not sure how tightly coupled the function selection is between them right now as the downsample version of this is.. brand new and looks weird.